// @oitq/core/lib/static.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultBotOptions = exports.defaultAppOptions = exports.dir = void 0;
const path = __importStar(require("path"));
const os = __importStar(require("os"));
exports.dir = path.join(os.homedir(), ".oitq");
exports.defaultAppOptions = {
    prefix: () => '',
    bots: [],
    maxListeners: 50,
    plugins: {},
    minSimilarity: 0.4,
    token: '',
    dir: process.cwd(),
    delay: {
        prompt: 60000
    }
};
exports.defaultBotOptions = {
    type: 'qrcode',
    admins: [],
    master: 1659488338,
    config: {
        platform: 5,
        data_dir: path.join(exports.dir, 'data')
    }
};
